packages/jupyter-ai/jupyter_ai/chat_handlers/base.py

```diff
@@ -114,7 +115,28 @@ async def handle_exc(self, e: Exception, message: HumanChatMessage):
         implementation is provided, however chat handlers (subclasses) should
         implement this method to provide a more helpful error response.
         """
-        await self._default_handle_exc(e, message)
+        self.log.error(e)
+        if self.is_api_key_exc(e):
+            self.handle_api_key_exc(e, message)
+        else:
+            await self._default_handle_exc(e, message)
+
+    def is_api_key_exc(self, e: Exception):
+        """
+        Checks if the exception is an default model API key exception.
+        """
+        return (
+            OpenAIErrorUtil.is_api_key_exc(e)
+            or AI21ErrorUtil.is_api_key_exc(e)
+            or AnthropicErrorUtil.is_api_key_exc(e)
```

Comment:
This looks like it could be difficult to extend to cover third-party model providers. What is the rationale for not adding `is_api_key_exc` on BaseProvider (and iterating all providers here)?

Reply:
That's a really good point. Encapsulating `is_api_key_exc` in `BaseProvider` would allow us to just call `is_api_key_exc` of the current provider which is much cleaner, less coupled, and easier to extend. Thank you @krassowski.